File: LeetCode/easy/Code167.js
/**
 * 本题介绍：
 * 给定一个已按照 非递减顺序排列  的整数数组 numbers ，请你从数组中找出两个数满足相加之和等于目标数 target 。
函数应该以长度为 2 的整数数组的形式返回这两个数的下标值。numbers 的下标 从 1 开始计数 ，所以答案数组应当满足 1 <= answer[0] < answer[1] <= numbers.length 。
你可以假设每个输入 只对应唯一的答案 ，而且你 不可以 重复使用相同的元素。
示例：
输入：numbers = [2,7,11,15], target = 9
输出：[1,2]
解释：2 与 7 之和等于目标数 9 。因此 index1 = 1, index2 = 2 。
 * @author dyp
 * javascript
 **/


/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
 var twoSum = function(numbers, target) {
    var len = numbers.length;
    for(var i=0;i<len-1;i++){
        /**
      * 解题思路 
      * 不完美，使用了双层for循环，时间复杂度太大
      */
        temp = numbers[i];
        for(var j = i+1;j<len;j++)
        {
            if(temp + numbers[j] == target)
                return [i+1,j+1]
        }
    }

};

/**
* 比较优秀的解题思路 
* 方法二：双指针
1. 定义left、right
2. 遍历过程若nums[left] + nums[right] > target，太大了，要小一点，则right左移一位
3. 遍历过程若nums[left] + nums[right] < target，太小了，要大一点，则left右移一位
4. 其余的情况就是nums[left] + nums[right] == target，return [left + 1, right + 1]即可
5. 注意题目的index从1开始，所以return时+1即可
*/
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
 var twoSum = function(numbers, target) {
    let [letf,right] = [0,numbers.length-1]
    while(letf<right){
        if(numbers[letf] + numbers[right] > target){
            right --;
        }else if(numbers[letf] + numbers[right] < target){
            letf++;
        }else{
            return [letf + 1,right + 1];
        }
    }

};

File: zyx/Leetcode/src/easy/Code167.java
package easy;

/**
 * Code167：给定一个已按照 非递减顺序排列  的整数数组 numbers ，请你从数组中找出两个数满足相加之和等于目标数 target 。
 * 函数应该以长度为 2 的整数数组的形式返回这两个数的下标值。numbers 的下标 从 1 开始计数 ，所以答案数组应当满足 1 <= answer[0] < answer[1] <= numbers.length 。
 * 你可以假设每个输入 只对应唯一的答案 ，而且你 不可以 重复使用相同的元素。
 *
 * @author tony
 */
public class Code167 {


    /**
     * 方法一：双指针解题
     * 定义low指向数组的首位，定义high指向数组的末尾
     * 每次计算两数和与目标值的大小，若大于，则右侧指针减少1，若小于，则左侧指针增加1
     * @param numbers
     * @param target
     * @return
     */
    public int[] twoSum(int[] numbers, int target) {
        //定义双指针
        int low = 0 , high = numbers.length - 1;
        //循环条件，low小于high
        while (low < high){
            //求和
            int sum = numbers[low] + numbers[high];
            //相等，则表示找到
            if (sum == target) {
                //+1是题目中的例子要求
                return new int[]{low+1, high+1};
                //小于的时候，右侧减1
            }else if(sum > target){
                --high;
                //大于的时候，左侧加1
            }else {
                ++low;
            }
        }
        return new int[]{-1,-1};
    }


    /**
     * 方法二：双指针+二分查找
     * @param numbers
     * @param target
     * @return
     */
    public int[] twoSum2(int[] numbers, int target) {
        int low = 0 , high = numbers.length - 1;
        while (low < high){
            int mid = low + (high - low) / 2;
            if(numbers[low] + numbers[high] > target){
                --high;
            }else if (numbers[low] + numbers[high] < target) {
                ++low;
            }else if (numbers[low] + numbers[mid] > target) {
                high = mid - 1;
            }else if (numbers[mid] + numbers[high] < target) {
                low = mid + 1;
            }else {
                return new int[]{low+1, high+1};
            }
        }
        return new int[]{-1,-1};
    }

}
